docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,69 @@
+# Heapster Push Metrics
+
+## Overview and Motivation
+
+Currently, Heapster supports pulling metrics from kubelet, and defines an
+interface for pulling from other sources.  However, in certain cases, it is
+more useful to be able to have services push metrics into Heapster, instead
+of having Heapster pull metrics.
+
+For instance, supporting push metrics makes it easy for cluster admins to add
+custom metrics into Heapster with relatively minimal effort: they can simply
+write a program or script which collects and processes the information, and
+then add a recurring Job or cron job which pushes the metrics (instead of
+having to write a source for Heapster).  This also enables existing tooling
+designed with the push model in mind to be adapted to provide metrics through
+Heapster.
+
+## Design
+
+### API
+
+Metrics added through push metrics will be available as normal through the
+Heapster model.
+
+To add metrics, metrics producers will `POST` new metrics to
+`/api/v1/push/{format}` in the format specified by `{format}`.  Producers will
+be authenticated similarly to the way Heapster currently authenticates clients:
+producers will present a certificate signed by a CA, and Heapster will be
+configurable to only allow certain names to push metrics.
+
+Pushed metrics will then be additionally scoped by producer name, to avoid
+conflicts.
+
+### Metrics Format
+
+The underlying design can support multiple format "backends".  Initially, only
+the prometheus text format will be supported at `/api/v1/push/prometheus`.  The
```

Comment:
This path does not follow the Prometheus Pushgateway API, and thus a Prometheus client library can't be used to push to this. If you want to offer this API then it should end with `/metrics/job/jobvalue/label1/value1/label2/value2` etc. which will be supported by the various existing Prometheus client libraries.

The Go client will push protobuf rather than text, so you should choose the parser based on the Content-Type header. The Go client has the required APIs to handle this for you.